src/ORM/EagerLoadedDataList.php

```diff
@@ -0,0 +1,893 @@
+     */
+    public function getExtraData($componentName, int|string $itemID): array
+    {
+        if (!in_array(get_class($this->datalist), [ManyManyList::class, ManyManyThroughList::class])) {
+            throw new BadMethodCallException('Cannot have extra fields on this list type');
+        }
+
+        // Allow string IDs for compatability with ManyManyList
+        if (!is_numeric($itemID)) {
+            throw new InvalidArgumentException('$itemID must be an integer or numeric string');
+        }
+
+        $itemID = (int)$itemID;
+        $rows = $this->getFinalisedRows();
+
+        // Skip if no extrafields or record not in this list
+        if (empty($this->extraFields) || !array_key_exists($itemID, $rows)) {
+            return [];
+        }
+
+        $result = [];
+        foreach ($rows[$itemID] as $fieldName => $value) {
+            if (array_key_exists($fieldName, $this->extraFields)) {
+                $result[$fieldName] = $value;
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * Gets the extra fields included in the relationship.
+     *
+     * @return array a map of field names to types
+     * @throws BadMethodCallException if the relation type for this list is not many_many
+     */
+    public function getExtraFields(): array
+    {
+        if (!in_array(get_class($this->datalist), [ManyManyList::class, ManyManyThroughList::class])) {
+            throw new BadMethodCallException('Cannot have extra fields on this list type');
+        }
+        return $this->extraFields;
+    }
+
+    private function setDataObjectEagerLoadedData(int $id, DataObject $item): void
+    {
+        if (array_key_exists($id, $this->eagerLoadedData)) {
+            foreach ($this->eagerLoadedData[$id] as $relation => $data) {
+                $item->setEagerLoadedData($relation, $data);
+            }
+        }
+    }
+
+    /**
+     * Gets the final rows for this list after applying all transformations.
+     * Currently only limit is applied lazily, but others could be done this was as well.
+     */
+    private function getFinalisedRows(): array
+    {
+        return $this->doLimit($this->doSort($this->rows));
+    }
+
+    private function doLimit(array $rows): array
+    {
+        list($length, $offset) = $this->limitOffset;
+
+        // If the limit is 0, return an empty list.
+        if ($length === 0) {
+            return [];
+        }
+
+        return array_slice($rows, $offset, $length, true);
+    }
+
+    private function doSort(array $rows): array
+    {
+        // Do nothing if there's no defined sort order.
+        if (empty($this->sort)) {
+            return $rows;
+        }
+
+        if ($this->sort === 'shuffle') {
+            ArrayLib::shuffleAssociative($rows);
+            return $rows;
+        }
+
+        uasort($rows, function (array $row, array $other): int {
+            $compared = 0;
+            foreach ($this->sort as $column => $direction) {
```

Comment:
should possibly use one of the natsort (natsort, natcasesort, strnatcmp, strnatcasecmp) methods here rather than straight <=> so that numbers get sorted better

https://www.php.net/manual/en/function.natsort.php

Reply:
It turns out that `<=>` already does natural sorting - but the database _doesn't_ do natural sorting on numeric strings. So I've had to treat numbers differently than numeric strings to match db behaviour.